lib/kubernetes-deploy/render_task.rb

```diff
@@ -0,0 +1,119 @@
+      else
+        only_filenames
+      end
+
+      validate_configuration(filenames)
+      render_filenames(stream, filenames)
+
+      @logger.summary.add_action("Successfully rendered #{filenames.size} template(s)")
+      @logger.print_summary(:success)
+    rescue KubernetesDeploy::FatalDeploymentError
+      @logger.print_summary(:failure)
+      raise
+    end
+
+    private
+
+    def render_filenames(stream, filenames)
+      exceptions = []
+      @logger.phase_heading("Rendering template(s)")
+
+      filenames.each do |filename|
+        begin
+          render_filename(filename, stream)
+        rescue KubernetesDeploy::InvalidTemplateError => exception
+          exceptions << exception
+          log_invalid_template(exception)
+        end
+      end
+
+      unless exceptions.empty?
+        raise exceptions[0]
+      end
+    end
+
+    def render_filename(filename, stream)
+      @logger.info("Rendering #{File.basename(filename)} ...")
+      file_content = File.read(File.join(@template_dir, filename))
+      rendered_content = @renderer.render_template(filename, file_content)
+      YAML.load_stream(rendered_content, "<rendered> #{filename}") do |doc|
+        stream.puts YAML.dump(doc)
+      end
+      @logger.info("Rendered #{File.basename(filename)}")
+    rescue Psych::SyntaxError => exception
+      raise InvalidTemplateError.new("Template is not valid YAML. #{exception.message}", filename: filename)
+    end
+
+    def validate_configuration(filenames)
+      @logger.info("Validating configuration")
+      errors = []
+
+      if filenames.empty?
+        errors << "no templates found in template dir #{@template_dir}"
+      end
+
+      absolute_template_dir = File.expand_path(@template_dir)
+
+      filenames.each do |filename|
+        absolute_file = File.expand_path(File.join(@template_dir, filename))
+        if !File.exist?(absolute_file)
+          errors << "Filename \"#{absolute_file}\" could not be found"
+        elsif !File.file?(absolute_file)
+          errors << "Filename \"#{absolute_file}\" is not a file"
+        elsif !absolute_file.start_with?(absolute_template_dir)
+          errors << "Filename \"#{absolute_file}\" is outside the template directory," \
+          " which was resolved as #{absolute_template_dir}"
+        end
+      end
+
+      unless errors.empty?
+        @logger.summary.add_action("Configuration invalid")
+        @logger.summary.add_paragraph(errors.map { |err| "- #{err}" }.join("\n"))
+        raise KubernetesDeploy::TaskConfigurationError, "Configuration invalid: #{errors.join(', ')}"
+      end
+    end
+
+    def log_invalid_template(exception)
+      @logger.error("Failed to render #{exception.filename}")
+
+      debug_msg = ColorizedString.new("Invalid template: #{exception.filename}\n").red
+      debug_msg += "> Error message: #{exception}"
+      if exception.content
+        debug_msg += "> Template content:\n#{exception.content}"
```

Comment:
I think we need a newline at the beginning of this as well. It's getting stuck on the end of the exception:

![image](https://user-images.githubusercontent.com/4789493/49970823-89ae9800-fefa-11e8-9d97-3c736cc3ba1e.png)